### PR TITLE
Add file temperature related counter and bytes stats to and io_stats

### PR DESCRIPTION
**Pull request by zhichao-cao**
For tiered storage project, we need to know the block read count and read bytes of files with different temperature. Add FileIOByTemperature to IOStatsContext and collect the bytes read and read count from different temperature files through the RandomAccessFileReader.

test plan: make check, add the testing cases

### CODE REVIEW

**Comment by zhichao-cao** [comment]
@siying The PR is updated accordingly.

File: include/rocksdb/iostats_context.h [hunk 1]
```diff
@@ -48,6 +68,8 @@ struct IOStatsContext {
   uint64_t cpu_write_nanos;
   // CPU time spent in read() and pread()
   uint64_t cpu_read_nanos;
+
+  FileIOByTemperature file_io_stats_by_temperature;
```
Comment (siying) [inline include/rocksdb/iostats_context.h:72]: What's the difference between the one here and the one in PerfContext?

Comment (siying) [reply include/rocksdb/iostats_context.h:72]: I guess the difference is bytes and count? Can we measure both in RandomAccessFileReader and put it here? In that way, we don't have to touch BlockBasedTableReader, so the logic would be simpler. Would that work?

Comment (zhichao-cao) [reply include/rocksdb/iostats_context.h:72]: Sure, I will move the counter to iostatus and measure them in RandomAccesssFuleReader::Read()

File: file/random_access_file_reader.cc [hunk 2]
```diff
@@ -22,6 +22,50 @@
 #include "util/rate_limiter.h"
 
 namespace ROCKSDB_NAMESPACE {
+inline void IOStatsAddBytesIfPositiveByTemperature(Temperature file_temperature,
```
Comment (siying) [inline file/random_access_file_reader.cc:25]: I think we can remove "if positive" to make the function name shorter.

Comment (siying) [reply file/random_access_file_reader.cc:25]: I'm a little bit worried about performance in this function. Although most of the time this path might involve I/O so performance might not matter, but some users might use something like mmap and I'm worried that we see a performance regression here.

I think to be safe, we might do an if for Temperature::kUnknown and skip everything. We might not need file_io_stats_by_temperature.unknown_file_bytes_read as we always know the total bytes and we can take difference.

By the way, it is not related to PR but IOSTATS_ADD_IF_POSITIVE() is a function I don't understand. Adding a thread local variable by 0 always feels cheaper than doing an if. Also value is almost never 0 so it's a waste. I suggest we just call IOSTATS_ADD() here.

**Review by siying — COMMENTED** [review]
I think we are almost there but I have some final implementation suggestions.

File: file/random_access_file_reader.cc (hunk 2 above)
Comment (siying) [reply file/random_access_file_reader.cc:25]: Let me create a PR to remove IOSTATS_ADD_IF_POSITIVE(). It doesn't make sense to me at all.

Comment (zhichao-cao) [reply file/random_access_file_reader.cc:25]: Thanks for the review. I will change to IOSTATS_ADD() and remove the `unknown` related counter and skip it at the beginning

**Pull request closed without merging**